File: C#/Get_Data/Transaction_Notification.cs
/*
* BluePay C#.NET Sample code.
*
* This code sample shows a very based approach
* on handling data that is posted to a script running
* a merchant's server after a transaction is processed
* through their BluePay gateway account.
*/

using System;
using BluePayLibrary;
using System.Collections.Specialized;
using System.Web;
using System.Net;
using System.IO;
using System.Text;

namespace GetData
{
    class TransactionNotification
    {
        public static void Main()
        {
            string accountID = "Merchant's Account ID Here";
            string secretKey = "Merchant's Secret Key Here";
            string mode = "TEST";
            
            BluePay tps = new BluePay
            (
                accountID,
                secretKey,
                mode
            );

            HttpListener listener = new HttpListener();
            string response = "";

            try
            {
                // Listen for incoming data
                listener.Start();
            }
            catch (HttpListenerException)
            {
                return;
            }
            while (listener.IsListening)
            {
                var context = listener.GetContext();
                var body = new StreamReader(context.Request.InputStream).ReadToEnd();

                byte[] b = Encoding.UTF8.GetBytes("ACK");

                // Return HTTP Status of 200 to BluePay
                context.Response.StatusCode = 200;
                context.Response.KeepAlive = false;
                context.Response.ContentLength64 = b.Length;

                var output = context.Response.OutputStream;
                output.Write(b, 0, b.Length);

                // Get Reponse
                using (StreamReader reader = new StreamReader(output))
                {
                    response = reader.ReadToEnd();
                }
                context.Response.Close();
            }
            listener.Close();
            NameValueCollection vals = HttpUtility.ParseQueryString(response);

            // Parse data into a NVP collection
            string transID = vals["trans_id"];
            string transStatus = vals["trans_stats"];
            string transType = vals["trans_type"];
            string amount = vals["amount"];
            string rebillID = vals["rebill_id"];
            string rebillAmount = vals["rebill_amount"];
            string rebillStatus = vals["rebill_status"];
            string tpsHashType = vals["TPS_HASH_TYPE"];
            string bpStamp = vals["BP_STAMP"];
            string bpStampDef = vals["BP_STAMP_DEF"];

            // calculate the expected BP_STAMP
            string bpStampString = "";
            string[] defSeparator = new string[] { " " };
            string[] bpStampFields = bpStampDef.Split(defSeparator, StringSplitOptions.RemoveEmptyEntries);
            foreach (string field in bpStampFields)
            {
                bpStampString += vals[field];
            }
            string expectedStamp = tps.GenerateTPS(bpStampString, tpsHashType);

            // Output data if the expected BP_STAMP matches the actual BP_STAMP
            if (expectedStamp == bpStamp) {
                Console.Write("Transaction ID: " + transID);
                Console.Write("Transaction Status: " + transStatus);
                Console.Write("Transaction Type: " + transType);
                Console.Write("Transaction Amount: " + amount);
                Console.Write("Rebill ID: " + rebillID);
                Console.Write("Rebill Amount: " + rebillAmount);
                Console.Write("Rebill Status: " + rebillStatus);
            } else {
                Console.Write("ERROR IN RECEIVING DATA FROM BLUEPAY");
            }
        }
    }
}

File: VB/vbnet/Get_Data/Transaction_Notification.vb
' *
' * Bluepay VB.NET Sample code.
' *
' * This code sample shows a very based approach
' * on handling data that is posted to a script running
' * a merchant's server after a transaction is processed
' * through their BluePay gateway account.
' *

 Imports System
 Imports vbnet.BPVB
 Imports System.IO
 Imports System.Collections.Specialized
 Imports System.Web
 Imports System.Net

 Namespace GetData

     Public Class TransactionNotification

        Public Sub New()
            MyBase.New()
        End Sub

        Public Shared Sub run()
            Dim accountID As String = "Merchant's Account ID Here"
            Dim secretKey As String = "Merchant's Secret Key Here"
            Dim mode As String = "TEST" 

            Dim tps As BluePay = New BluePay(
                accountID,
                secretKey,
                mode
            )

            Dim listener As HttpListener = New HttpListener()
            Dim prefix As String = "http://localhost:8080/" ' Change this to your listener URL (and optionally port)
            Dim responseString As String = ""

            Try
                'Listen for incoming data
                listener.Start()
            Catch ex As HttpListenerException
                Return
            End Try
            listener.Prefixes.Add(prefix)
            While (listener.IsListening)
                Dim context As HttpListenerContext = listener.GetContext()
                Dim response As HttpListenerResponse = context.Response
                Dim body As String = New StreamReader(context.Request.InputStream).ReadToEnd()

                ' Return HTTP Status of 200 to BluePay
                context.Response.StatusCode = 200
                context.Response.KeepAlive = False
                context.Response.ContentLength64 = body.Length

                ' Get Response
                Dim buffer() As Byte = System.Text.Encoding.UTF8.GetBytes(body)
                response.ContentLength64 = buffer.Length
                Dim output As System.IO.Stream = response.OutputStream
                output.Write(buffer, 0, buffer.Length)
                responseString = System.Text.Encoding.ASCII.GetString(buffer)
                context.Response.Close()

                ' Parse data into a NVP collection
                Dim vals As NameValueCollection = HttpUtility.ParseQueryString(responseString)
                Dim bpStampDef As String = vals("BP_STAMP_DEF")
                Dim tpsHashType As String = vals("TPS_HASH_TYPE")
                Dim bpStampVals As String
                Dim strArr() As String
                Dim i As Integer
                strArr = bpStampDef.Split(" ")
                For i = 0 To strArr.Length - 1
                    bpStampVals += vals(strArr(i))
                Next

                ' calculate the expected BP_STAMP
                Dim bpStamp As String = tps.generateTPS(bpStampVals, tpsHashType)

                ' Output data if the expected bp_stamp matches the actual BP_STAMP
                If bpStamp = vals("BP_STAMP") Then
                    Console.Write("Transaction ID: " + vals("trans_id"))
                    Console.Write("Transaction Status: " + vals("trans_stats"))
                    Console.Write("Transaction Type: " + vals("trans_type"))
                    Console.Write("Transaction Amount: " + vals("amount"))
                    Console.Write("Rebill ID: " + vals("rebill_id"))
                Else
                    Console.Write("ERROR IN RECEIVING DATA FROM BLUEPAY")
                End If

            End While
            listener.Close()
        End Sub
    End Class
 End Namespace
''